### PR TITLE
add numpy backend ops

### PR DESCRIPTION
**Pull request by danFromTelAviv**
**Summary**
Comparing to numpy operations allow us to know which backend fails in case of unexpected result (let's say the theano backend changes, the failure will only happen in the theano build on travis when comparing to numpy), and have a good ground truth for tests.

It's also good for documentation purposes, in case we wonder what the real definition of an op in the keras backend is.

**Related Issues**
**PR Overview**
- [ ] This PR requires new unit tests [y/n] (make sure tests are included)
- [ ] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

File: tests/keras/backend/backend_test.py [hunk 1]
```diff
@@ -300,8 +300,8 @@ def test_repeat_elements(self):
     def test_tile(self):
         shape = (3, 4)
         arr = np.arange(np.prod(shape)).reshape(shape)
-        check_single_tensor_operation('tile', arr, BACKENDS, n=[2, 1])
-        check_single_tensor_operation('tile', (2, 5), BACKENDS, n=[5, 2])
+        check_single_tensor_operation('tile', arr, WITH_NP, n=[2, 1])
```
Comment (gabrieldemarmiesse) [inline tests/keras/backend/backend_test.py]: #11205 already adds the `tile` op. So I don't think it's necessary to add it in this PR.

File: tests/keras/backend/reference_operations.py [hunk 2]
```diff
@@ -526,6 +528,164 @@ def resize_volumes(x, depth_factor, height_factor, width_factor, data_format):
...
+            idx = x_ndim + y_ndim - 3
+        else:
+            idx = x_ndim - 1
+        out = np.squeeze(out, tuple(range(idx, idx + diff)))
+    if ndim(out) == 1:
+        out = expand_dims(out, 1)
+    return out
+
+
+def local_conv1d(inputs, kernel, kernel_size, strides, data_format=None):
```
Comment (gabrieldemarmiesse) [inline tests/keras/backend/reference_operations.py]: Let's not add ops that we don't use in the tests (for the moment, it can be added in a later PR). Changing `reference_operations.py` is to enable better testing in `backend_tests.py`.

File: tests/keras/backend/reference_operations.py [hunk 3]
```diff
@@ -526,6 +528,164 @@ def resize_volumes(x, depth_factor, height_factor, width_factor, data_format):
...
+    # Shape: `(output_length, batch_size, filters)`.
+    output = batch_dot(x_aggregate, kernel)
+    return permute_dimensions(output, (1, 0, 2))
+
+
+def int_shape(x):
+    dims = x.shape
+    if dims is not None:
+        return dims
+    return None
```
Comment (gabrieldemarmiesse) [inline tests/keras/backend/reference_operations.py]: Isn't it equivalent to this?
```python
def int_shape(x):
    return x.shape
```

File: tests/keras/backend/backend_test.py (hunk 1 above)
Comment (danFromTelAviv) [reply tests/keras/backend/backend_test.py]: I removed my version

File: tests/keras/backend/reference_operations.py (hunk 2 above)
Comment (danFromTelAviv) [reply tests/keras/backend/reference_operations.py]: I removed local_conv1d and 2d , any other functions?

File: tests/keras/backend/reference_operations.py (hunk 3 above)
Comment (danFromTelAviv) [reply tests/keras/backend/reference_operations.py]: yep.

File: tests/keras/backend/reference_operations.py [hunk 4]
```diff
@@ -118,10 +120,10 @@ def pool(x, pool_size, strides, padding, data_format, pool_mode):
...
                 for (m, m1) in zip(range(pool_size[2]), range(-pool_size[2], 0)):
                     y.append(x[:,
-                               :,
-                               k:k1:strides[0],
-                               l:l1:strides[1],
-                               m:m1:strides[2]])
+                             :,
+                             k:k1:strides[0],
+                             l:l1:strides[1],
+                             m:m1:strides[2]])
```
Comment (gabrieldemarmiesse) [inline tests/keras/backend/reference_operations.py]: There seems to be a hook commit changing the indentation here. Can you revert?

File: tests/keras/backend/reference_operations.py [hunk 5]
```diff
@@ -526,6 +528,101 @@ def resize_volumes(x, depth_factor, height_factor, width_factor, data_format):
     return x
 
 
+def ndim(x):
```
Comment (gabrieldemarmiesse) [inline tests/keras/backend/reference_operations.py]: I think that numpy arrays have a ndim attribute that you could use.

File: tests/keras/backend/reference_operations.py [hunk 6]
```diff
@@ -526,6 +528,101 @@ def resize_volumes(x, depth_factor, height_factor, width_factor, data_format):
...
+    dims = x.shape
+    if dims is not None:
+        return len(dims)
+    return None
+
+
+def batch_dot(x, y, axes=None):
```
Comment (gabrieldemarmiesse) [inline tests/keras/backend/reference_operations.py]: Numpy has a `tensordot` function. I don't know much about it but can it help in our case?

File: tests/keras/backend/reference_operations.py [hunk 7]
```diff
@@ -526,6 +528,101 @@ def resize_volumes(x, depth_factor, height_factor, width_factor, data_format):
...
+def cast(x, dtype):
+    return x.astype(dtype)
+
+
+def zeros(shape, dtype=None, name=None):
+    if dtype is None:
+        dtype = floatx()
+    v = np.zeros(shape=shape, dtype=dtype)
+    if py_all(list(v.shape)):
+        return variable(v, dtype=dtype, name=name)
```
Comment (gabrieldemarmiesse) [inline tests/keras/backend/reference_operations.py]: Is this condition necessary?

File: tests/keras/backend/reference_operations.py (hunk 4 above)
Comment (danFromTelAviv) [reply tests/keras/backend/reference_operations.py]: I used ctl+alt+l in pycharm to automatically fix formatting - my apologies. I undid the indentation.

File: tests/keras/backend/reference_operations.py (hunk 5 above)
Comment (danFromTelAviv) [reply tests/keras/backend/reference_operations.py]: sure . replaced by return x.ndim

File: tests/keras/backend/reference_operations.py (hunk 6 above)
Comment (danFromTelAviv) [reply tests/keras/backend/reference_operations.py]: Indeed it is useful - if you look closely you will see I use it in the function. I tried to stick as closely to the tf backend implementation as possible to avoid unexpected bugs.

File: tests/keras/backend/reference_operations.py (hunk 7 above)
Comment (danFromTelAviv) [reply tests/keras/backend/reference_operations.py]: I tried to stick as closely to the tf backend implementation as possible to avoid unexpected bugs.

Comment (danFromTelAviv) [reply tests/keras/backend/reference_operations.py]: If you prefect simplicity I can remove the if py_all ....

Comment (gabrieldemarmiesse) [reply tests/keras/backend/reference_operations.py]: We don't need to follow tensorflow's code. The most important things are simplicity, readability and correctness. The tests are here for the correctness.

**Comment by danFromTelAviv** [comment]
looks like someone accepted a pull request where someone changed test_relu() and it no longer passes for multiple reasons.

**Comment by danFromTelAviv** [comment]
Looks like the CI is getting stuck sometimes and then fails on timeout. What can I do about that?

**Comment by danFromTelAviv** [comment]
still timing out... I'll try again tomorrow.

**Comment by danFromTelAviv** [comment]
still times out on : 
[gw0] [ 91%] PASSED tests/keras/wrappers/scikit_learn_test.py::test_regression_class_build_fn

**Comment by gabrieldemarmiesse** [comment]
This problem happens sometime, but it's the first time that it happens in a consistent manner. Thanks for your patience.

**Comment by danFromTelAviv** [comment]
> Thanks for your patience. I'll look into it.
It is a huge honor for me to add even a tiny piece of code into keras. I'm here as long as it takes to get this done. 

Maybe the fastest thing would be to try a new fork and add once functiona at a time and take the loss?

**Comment by gabrieldemarmiesse** [comment]
It would be nice if you could open a new PR. Maybe that will fix the travis issue. Also, in general, to avoid messy git history commits, you should not use your master branch for PRs. You should keep it to have a copy of the remote master branch locally. (lots of unrelated commits in this pr as you can see because of the merge).

What you should do:

1) Delete your fork (but keep a copy of the changes somewhere)
2) Create a fork of keras again
3) Add keras as the upstream repo `git remote add upstream https://github.com/keras-team/keras.git`
4) Update your master branch frequently by doing `git checkout master && git pull upstream master`
5) To make a PR create a new branch from your master `git checkout -b branch_for_PR_42`
6) If you want to update your PR branch, update master with `git checkout master && git pull upstream master ` first and then merge master into your branch `git checkout branch_for_PR_42 && git merge master`.

You'll be able to do multiple PR with one single fork and also avoid the many many commits added when merging like what happened here.

**Comment by danFromTelAviv** [comment]
yeah - I must admit this got a bit out of hand. I didn't expect it to be such a long PR so I didn't do it "by the book".

**Comment by gabrieldemarmiesse** [comment]
No worries :)

**Comment by gabrieldemarmiesse** [comment]
Closing in favor of a new, smaller and cleaner PR.
